node/test/Test/Pos/Block/Logic/VarSpec.hs

```diff
@@ -288,3 +303,97 @@ verifyBlockScenarioResult = \case
+applyThroughEpochSpec :: HasCoreConstants => Int -> Spec
+applyThroughEpochSpec afterCross = do
+    prop applyThroughEpochDesc (applyThroughEpochProp afterCross)
+  where
+    applyThroughEpochDesc =
+      "apply a sequence of blocks that spans through epochs (additional blocks after crossing: " ++
+      show afterCross ++ ")"
+
+applyThroughEpochProp :: HasCoreConstants => Int -> BlockProperty ()
+applyThroughEpochProp afterCross = do
+    scenario <- blockPropertyScenarioGen $ do
+        let
+            approachEpochEdge =
+                pathSequence mempty . OldestFirst . NE.fromList $
+                replicate (fromIntegral epochSlots - 1) "a"
+            crossEpochEdge =
+                pathSequence (NE.last $ getOldestFirst approachEpochEdge) $
+                OldestFirst . NE.fromList $
+                -- 2 blocks to ensure that we cross,
+                -- then some additional blocks
+                replicate (afterCross + 2) "x"
+        emitBlockApply BlockApplySuccess approachEpochEdge
+        emitBlockApply BlockApplySuccess crossEpochEdge
+    runBlockScenarioAndVerify scenario
+
+----------------------------------------------------------------------------
+-- Forks
+----------------------------------------------------------------------------
+
+singleForkSpec :: HasCoreConstants => ForkDepth -> Spec
+singleForkSpec fd = do
+    prop singleForkDesc (singleForkProp fd)
+  where
+    singleForkDesc =
+      "a blockchain of length q<=(9.5*k) blocks can switch to a fork " <>
+      "of length j>i with a common prefix i, rollback depth d=q-i"
+
+singleForkProp :: HasCoreConstants => ForkDepth -> BlockProperty ()
+singleForkProp fd = do
+    scenario <- blockPropertyScenarioGen $ genSingleFork fd
+    runBlockScenarioAndVerify scenario
+
+data ForkDepth = ForkShort | ForkMedium | ForkDeep
+
+genSingleFork :: forall g m. (HasCoreConstants, RandomGen g, Monad m) => ForkDepth -> BlockEventGenT g m ()
+genSingleFork fd = do
+    let k = fromIntegral blkSecurityParam :: Int
+    -- 'd' is how deeply in the chain the fork starts. In other words, it's how many
+    -- blocks we're going to rollback (therefore must be >1).
+    d <- getRandomR $ case fd of
+        ForkShort  -> (1, if k > 1 then k-1 else 1)
+        ForkMedium -> (if k > 2 then k - 2 else 1, k+2)
+        ForkDeep   -> (k+1, div (k*3) 2 + 1)
+    -- the depth must be <=k for a successful rollback.
+    let expectSuccess = d <= k
+    -- original blockchain max index q<(9.5*k)
+    q <- getRandomR (d+1, 9 * k + div k 2)
+    let
+        -- max index of the common prefix. i>0 because d<q
+        i = q-d
+    -- fork blockchain max index j>i. the upper bound is arbitrary.
+    -- dj=j-i
+    dj <- getRandomR (1, d*2)
+    -- now we can generate paths:
+    --
+    -- B0 - B1 - B2 - B3 - B4 - B5 - B6 - B7
+    --              \
+    --                C3 - C4 - C5 - C6
+    --
+    -- in this example, q=7, d=5, i=2, dj=4
```

Comment:
It would be good to also write that `B0` is 0-th genesis block which shouldn't be generated. I didn't understand what was going on, wrote a big wall of text and only then realized it.